lib/Runtime/Library/JavascriptGenerator.cpp

```diff
@@ -247,6 +262,419 @@ namespace Js
+    }
+
+    Var JavascriptGenerator::EntryAsyncThrow(RecyclableObject* function, CallInfo callInfo, ...)
+    {
+        PROBE_STACK(function->GetScriptContext(), Js::Constants::MinStackDefault);
+
+        ARGUMENTS(args, callInfo);
+        ScriptContext* scriptContext = function->GetScriptContext();
+        JavascriptLibrary* library = scriptContext->GetLibrary();
+
+        AUTO_TAG_NATIVE_LIBRARY_ENTRY(function, callInfo, _u("AsyncGenerator.prototype.throw"));
+
+        Var input = args.Info.Count > 1 ? args[1] : library->GetUndefined();
+
+        return JavascriptGenerator::AsyncGeneratorEnqueue(args[0], scriptContext, input,
+            RecyclerNew(scriptContext->GetRecycler(), JavascriptExceptionObject, input, scriptContext, nullptr), _u("AsyncGenerator.prototype.throw"));
+    }
+
+
+    Var JavascriptGenerator::EntryAsyncGeneratorAwaitReject(RecyclableObject* function, CallInfo callInfo, ...)
+    {
+        ScriptContext* scriptContext = function->GetScriptContext();
+        PROBE_STACK(scriptContext, Js::Constants::MinStackDefault);
+        ARGUMENTS(args, callInfo);
+        Assert(!(callInfo.Flags & CallFlags_New));
+        AssertOrFailFastMsg(args.Info.Count > 1, "Should never call EntryAsyncGeneratorAwait without a parameter");
+
+        AsyncGeneratorNextProcessor* resumeNextReturnProcessor = VarTo<AsyncGeneratorNextProcessor>(function);
+
+        ResumeYieldData yieldData(args[1], RecyclerNew(scriptContext->GetRecycler(), JavascriptExceptionObject, args[1], scriptContext, nullptr));
+        resumeNextReturnProcessor->GetGenerator()->CallAsyncGenerator(&yieldData);
+        return scriptContext->GetLibrary()->GetUndefined();
+    }
+
+    Var JavascriptGenerator::EntryAsyncGeneratorAwaitRevolve(RecyclableObject* function, CallInfo callInfo, ...)
+    {
+        ScriptContext* scriptContext = function->GetScriptContext();
+        PROBE_STACK(scriptContext, Js::Constants::MinStackDefault);
+        ARGUMENTS(args, callInfo);
+        Assert(!(callInfo.Flags & CallFlags_New));
+        AssertOrFailFastMsg(args.Info.Count > 1, "Should never call EntryAsyncGeneratorAwait without a parameter");
+
+        AsyncGeneratorNextProcessor* resumeNextReturnProcessor = VarTo<AsyncGeneratorNextProcessor>(function);
+
+        ResumeYieldData yieldData(args.Values[1], nullptr);
+        resumeNextReturnProcessor->GetGenerator()->CallAsyncGenerator(&yieldData);
+        return scriptContext->GetLibrary()->GetUndefined();
+    }
+
+    void JavascriptGenerator::CallAsyncGenerator(ResumeYieldData* yieldData)
+    {
+        ScriptContext* scriptContext = this->GetScriptContext();
+        Var result = nullptr;
+        JavascriptExceptionObject *exception = nullptr;
+        yieldData->generator = this;
+        AssertMsg(isAsync, "Should not call CallAsyncGenerator on a non-async generator");
```

Comment:
Can this be AssertOrFailFast instead, just to forestall a type confusion security issue if a mistake is made in a caller somewhere down the road?

Reply:
In most cases the caller checks this so making it a FailFast means it's checked twice - but agree that getting this wrong would be an easy way to introduce a security bug in the future. Will update.